raft/raft.go

```diff
@@ -611,6 +611,11 @@ func (r *raft) becomePreCandidate() {
 	// but doesn't change anything else. In particular it does not increase
 	// r.Term or change r.Vote.
 	r.step = stepCandidate
+	if r.checkQuorum {
+		// If network partition happen, and leader is in minority partition,
+		// it will step down, become follower without notifying others.
+		r.lead = None
```

Comment:
Why does this change matter?

It doesn't seem that strange to me that other nodes would still consider node A to be the leader when it has stepped down due to CheckQuorum.

Reply:
As described in thesis section 6.2, Routing requests to the leader,  
```
Raft must also prevent stale leadership information from delaying client requests indefinitely.
Leadership information can become stale all across the system, in leaders, followers, and clients:

• Leaders: A server might be in the leader state, but if it isn’t the current leader, it could be
needlessly delaying client requests. For example, suppose a leader is partitioned from the
rest of the cluster, but it can still communicate with a particular client. Without additional
mechanism, it could delay a request from that client forever, being unable to replicate a log
entry to any other servers. Meanwhile, there might be another leader of a newer term that is
able to communicate with a majority of the cluster and would be able to commit the client’s
request. Thus, `a leader in Raft steps down if an election timeout elapses without a successful
round of heartbeats to a majority of its cluster; this allows clients to retry their requests with
another server.`

• Followers: Followers keep track of the leader’s identity so that they can redirect or proxy
clients. `They must discard this information when starting a new election or when the term
changes.` Otherwise, they might needlessly delay clients (for example, it would be possible
for two servers to redirect to each other, placing clients in an infinite loop).
```
followers need discard leader information when starting a new election or when the term changes.

if not,  in the previous scenario,  partition [A, B] has A with lead = None, B with lead = A, 
if client request servers with alphabetical order, it could delay forever.